### PR TITLE
parser: fix subquery hint panic (#8781)

### PR DESCRIPTION
**Pull request by lysu**
cherry pick #8781, refer https://github.com/pingcap/parser/pull/101's branch

please don't merge until https://github.com/pingcap/parser/pull/101 merged~

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8800)

### CODE REVIEW

**Comment by lysu** [comment]
/run-all-tests tidb-test=release-2.1 tikv=release-2.1 pd=release-2.1

**Comment by zhouqiang-cl** [comment]
/run-all-tests tidb-test=release-2.1 tikv=release-2.1 pd=release-2.1

File: go.mod [hunk 1]
```diff
@@ -81,4 +81,4 @@ require (
 	gopkg.in/yaml.v2 v2.2.1 // indirect
 )
 
-replace github.com/pingcap/parser => github.com/lysu/parser v0.0.0-20181221063414-a86a9ff7a7de
+replace github.com/pingcap/parser => github.com/lysu/parser v0.0.0-20181225044558-8fb64d4405d8
```
Comment (zimulala) [inline go.mod]: I merged pingcap/parser#101. So do we need to update this?

**Comment by zhouqiang-cl** [comment]
/run-all-tests tidb-test=release-2.1 tikv=release-2.1 pd=release-2.1

File: go.mod (hunk 1 above)
Comment (lysu) [reply go.mod]: done~thx

**Comment by lysu** [comment]
/run-all-tests tidb-test=release-2.1 tikv=release-2.1 pd=release-2.1

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**